### PR TITLE
TASK: Fusion performance optimization (get configuration)

### PR DESCRIPTION
**Pull request by hlubek**
**What I did**

* Refactored the configuration processing in `Runtime` to a separate `RuntimeConfiguration` class for easier testing and separation of concerns
* Optimized getting the configuration for a Fusion path, since this is a very "hot" part of the code that is used for every evaluation

**How I did it**

* Measured and verified performance boost with a large production Neos project
* Performance boost might only be measurable if a lot of Fusion is evaluated (against the total overhead)

**How to verify it**

* See new unit test

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

Resolves: #2796

### CODE REVIEW

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,235 @@
+<?php
+
```
Comment (daniellienert) [inline Neos.Fusion/Classes/Core/RuntimeConfiguration.php]:
```suggestion
declare(strict_types=1);
```

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * The parsed Fusion configuration
+     *
+     * @var array
+     */
+    private $fusionConfiguration;
+
+    /**
+     * @var array
```
Comment (daniellienert) [inline Neos.Fusion/Classes/Core/RuntimeConfiguration.php]:
```suggestion
     * @var string[]
```

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * Matches the current path segment and prepares the configuration.
+     *
+     * @param string $pathPart
+     * @param array $previousConfiguration
+     * @param array $currentPrototypeDefinitions
+     * @return array
+     * @throws Exception
+     */
+    private function matchCurrentPathPart($pathPart, $previousConfiguration, &$currentPrototypeDefinitions)
```
Comment (daniellienert) [inline Neos.Fusion/Classes/Core/RuntimeConfiguration.php]:
```suggestion
    private function matchCurrentPathPart(string $pathPart, array $previousConfiguration, arrax &$currentPrototypeDefinitions): array
```

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+    /**
+     * Merges the prototype chain into the configuration.
+     *
+     * @param array $configuration
+     * @param array $currentPrototypeDefinitions
+     * @return array
+     * @throws Exception
+     */
+    private function mergePrototypesWithConfigurationForPathSegment($configuration, &$currentPrototypeDefinitions)
```
Comment (daniellienert) [inline Neos.Fusion/Classes/Core/RuntimeConfiguration.php]:
```suggestion
    private function mergePrototypesWithConfigurationForPathSegment(array $configuration, array &$currentPrototypeDefinitions): array
```

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php (hunk 2 above)
Comment (hlubek) [reply Neos.Fusion/Classes/Core/RuntimeConfiguration.php]: Well, `pathCache` is actually an array of arrays indexed by string... Dunno how to type that in PHP ;)

**Comment by bwaidelich** [comment]
I added an issue for this in order to simplify release management and to collect higher-level details for the release notes: #2796

**Comment by davidspiola** [comment]
@daniellienert if you find the time to reviews this (just do make sure your changes are reflected) unitl tonight, it would be very helpfull. I'd like to merge it tonight. thx

File: Neos.Fusion/Classes/Core/Runtime.php [hunk 5]
```diff
@@ -110,7 +110,7 @@ class Runtime
     /**
      * @var array
```
Comment (kitsunet) [inline Neos.Fusion/Classes/Core/Runtime.php]:
```suggestion
     * @var RuntimeConfiguration
```

File: Neos.Fusion/Classes/Core/Runtime.php [hunk 6]
```diff
@@ -161,21 +145,9 @@ class Runtime
      */
     public function __construct(array $fusionConfiguration, ControllerContext $controllerContext)
     {
-        $this->fusionConfiguration = $fusionConfiguration;
+        $this->runtimeConfiguration = new RuntimeConfiguration($fusionConfiguration);
```
Comment (kitsunet) [inline Neos.Fusion/Classes/Core/Runtime.php]: Shouldn't this happen in a way that makes it easier to exchange the implementation for testing? I thought that was the general idea (together with performance) for this change. Like a factory for example?

File: Neos.Fusion/Classes/Core/RuntimeConfiguration.php [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        // Fast path if complete Fusion path is in configuration cache
+        if (isset($this->pathCache[$fusionPath])) {
+            return $this->pathCache[$fusionPath]['c'];
+        }
+
+        // Find longest prefix of path that already is in path cache
+        $pathUntilNow = '';
+        $fusionPathLength = strlen($fusionPath);
+        $offset = $fusionPathLength;
+        while (($offset = strrpos($fusionPath, '/', -($fusionPathLength - $offset + 1))) != false) {
```
Comment (kitsunet) [inline Neos.Fusion/Classes/Core/RuntimeConfiguration.php]: I would naturally rather do an explode here, and peel elements off the end in the loop, that seems easier to understand than all these string operations and is probably less error prone. But fine either way I guess.

File: Neos.Fusion/Classes/Core/Runtime.php (hunk 6 above)
Comment (hlubek) [reply Neos.Fusion/Classes/Core/Runtime.php]: Well, right now it was only meant for easier testing of the RuntimeConfiguration itself and better encapsulation of the configuration handling code.

Ideally a `RuntimeConfiguration` value should be passed as a dependency in the constructor, but that would change the signature.

**Pull request merged**